### PR TITLE
[Exclusivity] Make the performance inliner aware of exclusivity checks

### PR DESCRIPTION
**Pull request by shajrawi**
radar rdar://problem/45283321

### CODE REVIEW

**Comment by shajrawi** [comment]
@swift-ci Please test

**Comment by shajrawi** [comment]
@eeckstein Can you please review?

**Comment by shajrawi** [comment]
@swift-ci Please test

**Comment by shajrawi** [comment]
@swift-ci Please test

**Comment by shajrawi** [comment]
@swift-ci Please test

**Comment by shajrawi** [comment]
@swift-ci Please test

**Comment by shajrawi** [comment]
@swift-ci Please test

**Comment by shajrawi** [comment]
@swift-ci Please test

**Comment by shajrawi** [comment]
@swift-ci Please test Linux

File: lib/SILOptimizer/Transforms/PerformanceInliner.cpp [hunk 1]
```diff
@@ -407,6 +446,12 @@ bool SILPerformanceInliner::isProfitableToInline(
     }
   }
 
+  Weight EntryW = SPA->getWeight(CalleeEntry, CallerWeight);
```
Comment (eeckstein) [inline lib/SILOptimizer/Transforms/PerformanceInliner.cpp]: Why are you taking the weight of the entry block and not the weight of the block which contains the access(es)?

Comment (shajrawi) [reply lib/SILOptimizer/Transforms/PerformanceInliner.cpp]: When we visit the block that contains the accesses we might not have visited the entire function / further block might change our decision. by the time we traverse all blocks we 'lost' which block(s) contain accesses. but I agree this is less than ideal - will think about changing it

**Comment by shajrawi** [comment]
@eeckstein I changed the benefit calculation per our off-line discussion

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

**Comment by shajrawi** [comment]
@swift-ci Please smoke test and merge

**Comment by shajrawi** [comment]
@swift-ci Please smoke test and merge

**Comment by shajrawi** [comment]
@swift-ci Please smoke test and merge

**Review by eeckstein — APPROVED** [review]
lgtm

**Pull request merged**